Client/src/ChronologClientImpl.cpp

```diff
@@ -0,0 +1,195 @@
+
+#include <unistd.h>
+#include "ChronologClientImpl.h"
+#include "StorytellerClient.h"
+#include "city.h"
+
+std::mutex chronolog::ChronologClientImpl::chronologClientMutex;
+chronolog::ChronologClientImpl * chronolog::ChronologClientImpl::chronologClientImplInstance{nullptr};
+
+
+chronolog::ChronologClientImpl * chronolog::ChronologClientImpl::GetClientImplInstance(ChronoLog::ConfigurationManager const& confManager)
+{
+    std::lock_guard<std::mutex> lock_client(chronologClientMutex);
+    if( chronologClientImplInstance == nullptr)
+    {
+        chronologClientImplInstance = new ChronologClientImpl(confManager);
+    }
+
+    return chronologClientImplInstance;
+}
+
+chronolog::ChronologClientImpl::~ChronologClientImpl()
+{
+	//TODO: implement
+	delete storyteller;
+// shared_ptr..	delete rpcClient_;
+}
+
+int chronolog::ChronologClientImpl::Connect(const std::string &server_uri,
+                             std::string const& client_account,
+                             int &flags)
+{
+    std::lock_guard<std::mutex> lock_client(chronologClientMutex);
+    // if already connected return success
+    // if disconencting return failure....
+    if((clientState != UNKNOWN) && (clientState != SHUTTING_DOWN))
+    {  return CL_SUCCESS;  }  
+    else if (clientState == SHUTTING_DOWN) 
+    {  return CL_ERR_INVALID_ARG; }
+
+/*    if (client_id.empty()) {
+        char ip[16];
+        struct hostent *he = gethostbyname("localhost");
+        auto **addr_list = (struct in_addr **) he->h_addr_list;
+        strcpy(ip, inet_ntoa(*addr_list[0]));
+        std::string addr_str = ip + std::string(",") + std::to_string(getpid());
+        uint64_t client_id_hash = CityHash64(addr_str.c_str(), addr_str.size());
+        client_id = std::to_string(client_id_hash);
+    }
+    */
+    clientAccount = client_account;
+       //TODO: client_id must be assigned by the Visor server
+    uint64_t clock_offset = 0; //TODO: use this value with the clocksource...
+
+    auto return_code = rpcClient_->Connect(server_uri, clientAccount, flags, clock_offset);
+
+    if(return_code == CL_SUCCESS)
+    {    
+	clientState = CONNECTED; 
+        clientId = 7; //TODO: retain clientId provided by he Visor 
+    }
+    return return_code;
+}
+
+int chronolog::ChronologClientImpl::Disconnect( ) //const std::string &client_id, int &flags)
+{
+    std::lock_guard<std::mutex> lock_client(chronologClientMutex);
+
+    if((clientState == UNKNOWN) || (clientState == SHUTTING_DOWN))
+    {  return CL_ERR_NO_CONNECTION; }
+
+  //TODO : release all the acquired stories before asking to disconnect...
+
+    int flags=1;
+    auto return_code = rpcClient_->Disconnect(clientAccount , flags);
+    if(return_code == CL_SUCCESS)
+    {    
+	clientState = SHUTTING_DOWN; 
```

Comment:
does Disconnect leave something in the background before it returns? shutting_down is a bit ambiguous to me. To me, I think we should block Disconnect until everything is cleaned up for this particular connection. I would suggest something like DISCONNECTED or DOWN as the name after Disconnect.

Reply:
SHUTTING_DOWN is meant to be a final state , not transitional.  no new requests for data would be allowed if state = SHUTTING_DOWN